File: Sources/SomnioScene3D/ModelAssets.swift
import CoreGraphics
import Foundation
import ImageIO
import Logging
import RealityKit
import SomnioCore

/// Model-pack accessor for the RealityKit scene. A protocol over a `Bundle.main`-backed loader
/// with one structural constraint forced by RealityKit: `Entity(contentsOf:)` is async, but the
/// render surface that consumes these models is driven synchronously. Implementations therefore
/// load and cache every registry model in an async `prewarm()` (called from the sector-load path
/// before placement), and the per-request accessors are synchronous cache reads — a stem the
/// prewarm has not cached yet resolves `nil` (⇒ placeholder), never blocking and never trapping.
@MainActor public protocol ModelAssets {
    func prewarm() async
    func entity(forKind kind: WorldEntity.Kind, figure: Int16) -> Entity?
    func object(forID id: String) -> Entity?
    func floorMaterialTexture(forID id: String) -> TextureResource?
    func floorMaterialURL(forID id: String) -> URL?
}

/// Loads USDZ models from a runtime `Bundle` whose `Resources/` directory follows the subdirectory
/// layout produced by `Scripts/bundle-assets.sh` (`Models/`, `FloorMaterials/`), resolving the
/// sector format's semantic ids through the committed model registry. Loaded roots are cached as
/// prototypes and every accessor returns a fresh recursive clone — an `Entity` is a mutable
/// scene-graph node (transform, parent, animation state), so handing the same instance to
/// multiple world entities would share that state across them.
@MainActor public final class BundleMainModelAssets: ModelAssets {
    private static let logger = Logger(label: "de.tobiha.somnio.scene3d.assets")

    private let bundle: Bundle
    private let registry: ModelRegistry
    private var prototypes: [String: Entity] = [:]
    /// Negative cache: a stem whose URL or load already failed is not re-resolved on the next
    /// prewarm pass (no-asset-pack fallback stays cheap).
    private var prototypeMisses: Set<String> = []
    private var floorMaterialTextureCache: [String: TextureResource] = [:]
    private var floorMaterialMisses: Set<String> = []

    public init(bundle: Bundle = .main, registry: ModelRegistry? = nil) {
        self.bundle = bundle
        self.registry = registry ?? Self.loadBundledRegistry()
    }

    /// Resolves the committed model registry, degrading to the empty placeholder fallback with a
    /// logged error if the bundled JSON is (theoretically) missing or corrupt — the same graceful
    /// path as an absent model pack rather than a trap.
    private static func loadBundledRegistry() -> ModelRegistry {
        do {
            return try ModelRegistryCodec.bundledRegistry()
        } catch {
            logger.error("ModelRegistry.json missing or invalid; using placeholder fallback", metadata: ["error": "\(error)"])
            return .placeholderFallback
        }
    }

    /// Loads and caches every registry model prototype and floor-material texture. Call from the
    /// async sector-load path before entity placement; the synchronous accessors only ever read
    /// the caches this fills.
    public func prewarm() async {
        for entry in registry.allModelEntries
            where prototypes[entry.stem] == nil && !prototypeMisses.contains(entry.stem) {
            await loadPrototype(for: entry)
        }
        for material in registry.floorMaterials
            where floorMaterialTextureCache[material.stem] == nil && !floorMaterialMisses.contains(material.stem) {
            await loadFloorMaterialTexture(stem: material.stem)
        }
    }

    public func entity(forKind kind: WorldEntity.Kind, figure: Int16) -> Entity? {
        guard let entry = registry.model(forKind: kind, figure: figure) else { return nil }
        return clone(ofStem: entry.stem)
    }

    public func object(forID id: String) -> Entity? {
        guard let entry = registry.model(forObjectID: id) else { return nil }
        return clone(ofStem: entry.stem)
    }

    /// The dedicated floor-material texture for a sector's `floorMaterialID`, or `nil` when the id
    /// is unmapped or its texture was not warmed (absent/unloadable PNG) — the floor then falls
    /// back to a solid lit plane. A pure read of the cache `prewarm()` fills, like `object`/`entity`.
    public func floorMaterialTexture(forID id: String) -> TextureResource? {
        guard let stem = registry.floorMaterialStem(forID: id) else { return nil }
        return floorMaterialTextureCache[stem]
    }

    public func floorMaterialURL(forID id: String) -> URL? {
        guard let stem = registry.floorMaterialStem(forID: id) else { return nil }
        return bundle.url(forResource: stem, withExtension: "png", subdirectory: "FloorMaterials")
    }

    private func clone(ofStem stem: String) -> Entity? {
        prototypes[stem]?.clone(recursive: true)
    }

    private func loadPrototype(for entry: ModelEntry) async {
        guard let url = bundle.url(forResource: entry.stem, withExtension: "usdz", subdirectory: "Models") else {
            prototypeMisses.insert(entry.stem)
            Self.logger.info("model .usdz absent from bundle; rendering placeholder", metadata: ["stem": "\(entry.stem)"])
            return
        }
        do {
            // A rigged USDZ is an entity tree whose skeleton and animation library must survive,
            // so it loads as a root `Entity` hierarchy, never a flattened single `ModelEntity`.
            let prototype = try await Entity(contentsOf: url)
            let missing = ModelRegistry.missingClips(
                expected: entry.expectedClips,
                actual: RealityKitAnimationClips.names(in: prototype)
            )
            if !missing.isEmpty {
                Self.logger.error(
                    "model is missing expected animation clips; the glb→USDZ conversion likely collapsed its clip library",
                    metadata: ["stem": "\(entry.stem)", "missing": "\(missing.sorted().joined(separator: ", "))"]
                )
            }
            prototypes[entry.stem] = prototype
        } catch {
            prototypeMisses.insert(entry.stem)
            Self.logger.error("failed to load model", metadata: ["stem": "\(entry.stem)", "error": "\(error)"])
        }
    }

    private func loadFloorMaterialTexture(stem: String) async {
        guard let url = bundle.url(forResource: stem, withExtension: "png", subdirectory: "FloorMaterials") else {
            floorMaterialMisses.insert(stem)
            Self.logger.info("floor material .png absent from bundle; rendering the untextured floor", metadata: ["stem": "\(stem)"])
            return
        }
        guard let source = CGImageSourceCreateWithURL(url as CFURL, nil),
              let image = CGImageSourceCreateImageAtIndex(source, 0, nil),
              // The async initializer keeps the (documented main-actor-blocking) synchronous texture
              // upload off the first sector-load frame. Mipmaps matter here: the material tiles across
              // the whole sector floor, and un-mipped minification shimmers under the 3/4 camera.
              let texture = try? await TextureResource(
                  image: image,
                  options: .init(semantic: .color, mipmapsMode: .allocateAndGenerateAll)
              )
        else {
            floorMaterialMisses.insert(stem)
            Self.logger.error("floor material unloadable; rendering the untextured floor", metadata: ["stem": "\(stem)"])
            return
        }
        floorMaterialTextureCache[stem] = texture
    }
}

File: Web/src/scene/modelAssets.ts
import * as THREE from 'three'
import { GLTFLoader } from 'three/examples/jsm/loaders/GLTFLoader.js'
import { clone as cloneSkinned } from 'three/examples/jsm/utils/SkeletonUtils.js'
import {
  allModelEntries,
  floorMaterialStem,
  missingClips,
  modelForEntity,
  modelForObjectID,
} from '@/core/modelRegistry'
import type { ModelRegistry } from '@/core/modelRegistry'
import type { WorldEntityKind } from '@/core/worldEntity'

/**
 * Model-pack accessor, mirroring `BundleMainModelAssets`.
 *
 * The same structural constraint as the native loader applies for the same reason: loading is
 * async but the render surface is driven synchronously, so everything is warmed in `prewarm()`
 * and the per-request accessors are synchronous cache reads. A stem the prewarm has not cached
 * resolves `undefined` (a placeholder), never blocking and never throwing.
 *
 * Assets are `.glb` here rather than `.usdz`: the pipeline emits GLB *upstream* of the USDZ
 * conversion, so the browser consumes an earlier artifact of the same run rather than a new one.
 */
export interface ModelAssets {
  prewarm(): Promise<void>
  entity(kind: WorldEntityKind, figure: number): THREE.Object3D | undefined
  object(id: string): THREE.Object3D | undefined
  floorTexture(id: string): THREE.Texture | undefined
  clipsFor(root: THREE.Object3D): THREE.AnimationClip[]
}

interface Prototype {
  scene: THREE.Object3D
  clips: THREE.AnimationClip[]
}

export class HttpModelAssets implements ModelAssets {
  private readonly registry: ModelRegistry
  private readonly baseURL: string
  private readonly loader = new GLTFLoader()
  private readonly textureLoader = new THREE.TextureLoader()
  private readonly prototypes = new Map<string, Prototype>()
  private readonly textures = new Map<string, THREE.Texture>()
  /** Negative cache: a stem whose fetch already failed is not retried on the next prewarm. */
  private readonly misses = new Set<string>()
  private readonly clipsByRoot = new WeakMap<THREE.Object3D, THREE.AnimationClip[]>()

  constructor(registry: ModelRegistry, baseURL = '/assets') {
    this.registry = registry
    this.baseURL = baseURL.replace(/\/$/, '')
  }

  async prewarm(): Promise<void> {
    await Promise.all([
      ...allModelEntries(this.registry)
        .filter((entry) => !this.prototypes.has(entry.stem) && !this.misses.has(entry.stem))
        .map((entry) => this.loadPrototype(entry.stem, entry.expectedClips)),
      ...this.registry.floorMaterials
        .filter((rule) => !this.textures.has(rule.stem) && !this.misses.has(rule.stem))
        .map((rule) => this.loadFloorTexture(rule.stem)),
    ])
  }

  entity(kind: WorldEntityKind, figure: number): THREE.Object3D | undefined {
    const entry = modelForEntity(this.registry, kind, figure)
    return entry === undefined ? undefined : this.instantiate(entry.stem)
  }

  object(id: string): THREE.Object3D | undefined {
    const entry = modelForObjectID(this.registry, id)
    return entry === undefined ? undefined : this.instantiate(entry.stem)
  }

  floorTexture(id: string): THREE.Texture | undefined {
    const stem = floorMaterialStem(this.registry, id)
    return stem === undefined ? undefined : this.textures.get(stem)
  }

  clipsFor(root: THREE.Object3D): THREE.AnimationClip[] {
    return this.clipsByRoot.get(root) ?? []
  }

  /**
   * `SkeletonUtils.clone`, **not** `Object3D.clone`.
   *
   * A plain clone copies the mesh nodes but leaves every copy pointing at the prototype's
   * `Skeleton` and its bone objects. Two entities sharing one skeleton means the second one's
   * animation drives the first, so a room of peers all pose identically to whoever moved last.
   */
  private instantiate(stem: string): THREE.Object3D | undefined {
    const prototype = this.prototypes.get(stem)
    if (prototype === undefined) return undefined
    const copy = cloneSkinned(prototype.scene)
    this.clipsByRoot.set(copy, prototype.clips)
    return copy
  }

  private async loadPrototype(stem: string, expectedClips: readonly string[]): Promise<void> {
    try {
      const gltf = await this.loader.loadAsync(`${this.baseURL}/Models/${stem}.glb`)
      const clips = gltf.animations
      // Same clip-presence contract the conversion validator enforces: a naive export collapses
      // the clip library into one timeline, and the symptom downstream is a character that only
      // ever plays its first animation.
      const missing = missingClips(
        expectedClips,
        clips.map((clip) => clip.name)
      )
      if (missing.length > 0) {
        console.error(
          `model "${stem}" is missing expected animation clips: ${missing.join(', ')} — the export likely collapsed its clip library`
        )
      }
      this.prototypes.set(stem, { scene: gltf.scene, clips })
    } catch {
      this.misses.add(stem)
      console.warn(`model "${stem}" unavailable; rendering a placeholder`)
    }
  }

  private async loadFloorTexture(stem: string): Promise<void> {
    try {
      const texture = await this.textureLoader.loadAsync(`${this.baseURL}/FloorMaterials/${stem}.png`)
      texture.wrapS = THREE.RepeatWrapping
      texture.wrapT = THREE.RepeatWrapping
      texture.colorSpace = THREE.SRGBColorSpace
      // The material tiles across a whole sector floor; un-mipped minification shimmers badly
      // under the tilted camera, and anisotropy is what keeps it from sparkling at grazing angles.
      texture.generateMipmaps = true
      texture.minFilter = THREE.LinearMipmapLinearFilter
      texture.magFilter = THREE.LinearFilter
      texture.anisotropy = 8
      this.textures.set(stem, texture)
    } catch {
      this.misses.add(stem)
      console.warn(`floor material "${stem}" unavailable; rendering the untextured floor`)
    }
  }
}
